### PR TITLE
Transfer ownership of the collection

### PR DESCRIPTION
**Pull request by dastanfv**

### CODE REVIEW

File: ownership-chain/precompile/evolution-collection/contracts/EvolutionCollection.json [hunk 1]
```diff
@@ -1,131 +1,202 @@
 [
```
Comment (dastanfv) [inline ownership-chain/precompile/evolution-collection/contracts/EvolutionCollection.json:1]: sorry for the formatter

File: ownership-chain/pallets/laos-evolution/src/tests.rs [hunk 2]
```diff
@@ -74,6 +74,69 @@ fn create_collection_emits_event() {
...
+			LaosEvolution::transfer_ownership(
+				AccountId::from_str(ALICE).unwrap(),
+				AccountId::from_str(BOB).unwrap(),
+				0_u64
+			),
+			Error::<Test>::CollectionDoesNotExist
+		);
+
+		let collection_id = create_collection(ALICE);
+		let new_owner = AccountId::from_str(BOB).unwrap();
```
Comment (magecnion) [inline ownership-chain/pallets/laos-evolution/src/tests.rs:91]: I would separate this as a different unit test, so each test checks a single scenario

File: ownership-chain/pallets/laos-evolution/src/tests.rs [hunk 3]
```diff
@@ -74,6 +74,69 @@ fn create_collection_emits_event() {
...
+
+		// Non-owner cannot transfer collection
+		assert_noop!(
+			LaosEvolution::transfer_ownership(
+				AccountId::from_str(BOB).unwrap(),
+				new_owner,
+				collection_id
+			),
+			Error::<Test>::NoPermission
+		);
```
Comment (magecnion) [inline ownership-chain/pallets/laos-evolution/src/tests.rs:101]: I would separate this as a different unit test, so each test checks a single scenario

File: ownership-chain/precompile/evolution-collection/contracts/EvolutionCollection.sol [hunk 4]
```diff
@@ -21,9 +21,14 @@ interface EvolutionCollection {
...
-    event EvolvedWithExternalURI(
-        uint256 indexed _tokenId,
-        string _tokenURI
+    event EvolvedWithExternalURI(uint256 indexed _tokenId, string _tokenURI);
+
+    /// @notice Emitted when ownership of the collection changes
+    /// @param previousOwner the previous owner of the collection
+    /// @param newOwner the new owner of the collection
+    event OwnershipTransferred(
+        address indexed previousOwner,
```
Comment (magecnion) [inline ownership-chain/precompile/evolution-collection/contracts/EvolutionCollection.sol:30]:
```suggestion
        address indexed _previousOwner,
```

File: ownership-chain/precompile/evolution-collection/contracts/EvolutionCollection.sol [hunk 5]
```diff
@@ -21,9 +21,14 @@ interface EvolutionCollection {
...
+        address indexed newOwner
```
Comment (magecnion) [inline ownership-chain/precompile/evolution-collection/contracts/EvolutionCollection.sol:31]:
```suggestion
        address indexed _newOwner
```

File: ownership-chain/precompile/evolution-collection/contracts/EvolutionCollection.sol [hunk 6]
```diff
@@ -64,6 +69,11 @@ interface EvolutionCollection {
         string calldata _tokenURI
     ) external;
 
+    /// @notice Transfers ownership of the collection to a new account (`newOwner`).
+    /// @dev Call this function to transfer ownership of the collection, the caller must be the owner of the collection
+    /// @param newOwner The address to transfer ownership to.
+    function transferOwnership(address newOwner) external;
```
Comment (magecnion) [inline ownership-chain/precompile/evolution-collection/contracts/EvolutionCollection.sol:75]:
```suggestion
    function transferOwnership(address _newOwner) external;
```

File: ownership-chain/precompile/evolution-collection/src/lib.rs [hunk 7]
```diff
@@ -310,6 +319,37 @@ where
...
+
+		let to = input.read::<Address>()?.0;
+
+		LaosEvolution::<Runtime>::transfer_ownership(caller.into(), to.into(), collection_id)
+			.map_err(revert_dispatch_error)?;
+
+		let consumed_weight = LaosEvolutionWeights::<Runtime>::transfer_ownership();
+
+		LogsBuilder::new(context.address)
+			.log3(SELECTOR_LOG_OWNERSHIP_TRANSFERRED, caller, to, EvmDataWriter::new().build())
```
Comment (magecnion) [inline ownership-chain/precompile/evolution-collection/src/lib.rs:342]:
```suggestion
			.log2(SELECTOR_LOG_OWNERSHIP_TRANSFERRED, caller, to)
```

File: ownership-chain/precompile/evolution-collection-factory/src/tests.rs [hunk 8]
```diff
@@ -160,7 +160,7 @@ fn test_expected_cost_create_collection() {
 		// `create_collection` weight + insert account bytecode for the collection + log costs
 		precompiles()
 			.prepare_test(H160([1u8; 20]), H160(PRECOMPILE_ADDRESS), input)
-			.expect_cost(482420381) //[`WeightToGas`] set to 1:1 in mock
+			.expect_cost(486001381) //[`WeightToGas`] set to 1:1 in mock
 			.execute_some();
 	})
 }
```
Comment (magecnion) [inline ownership-chain/precompile/evolution-collection-factory/src/tests.rs:166]: what about adding `transfer_ownership_expected_cost`?

**Review by magecnion — APPROVED** [review]

**Review by asiniscalchi — APPROVED** [review]

**Pull request merged**